File: admin/src/main/java/com/project/admin/repository/GithubReleaseRepository.java
package com.project.admin.repository;

import com.project.admin.domain.entity.GithubRelease;
import com.project.admin.constant.Status;
import com.project.admin.repository.quertdsl.GithubReleaseRepositoryCustom;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;


public interface GithubReleaseRepository extends
        JpaRepository<GithubRelease, Long>,
        GithubReleaseRepositoryCustom {


    @Modifying
    @Query("UPDATE GithubRelease  g SET  g.status= :status where  g.id in :ids")
    void updateStatusByIds(@Param("ids") List<Long> ids,
            @Param("status") Status status);
}

File: api/src/main/kotlin/com/project/api/repository/GithubReleaseRepository.kt
package com.project.api.repository

import com.project.api.constants.TechStack
import com.project.api.entity.GithubRelease
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface GithubReleaseRepository : JpaRepository<GithubRelease, Long> {

    @Query(
        """
        SELECT g FROM GithubRelease g
        WHERE g.status = com.project.api.constants.Status.PUBLISHED
          AND g.draft = false
          AND g.prerelease = false
          AND (:#{#techStacks.size()} = 0 OR g.techStack IN :techStacks)
        ORDER BY g.publishedAt DESC
    """
    )
    fun findByTechStacks(
        @Param("techStacks") techStacks: Set<TechStack>,
        pageable: Pageable,
    ): Page<GithubRelease>


    @Query(
            """
    SELECT g FROM GithubRelease g
    WHERE g.status = com.project.api.constants.Status.PUBLISHED
      AND g.draft = false
      AND g.prerelease = false
      AND (:#{#techStacks.size()} = 0 OR g.techStack IN :techStacks)
      AND (
        :keyword IS NULL
        OR (:#{#matchedTechStacks.size()} > 0 AND g.techStack IN :matchedTechStacks)
        OR g.name LIKE CONCAT('%', :keyword, '%')
        OR g.tagName LIKE CONCAT('%', :keyword, '%')
      )
    ORDER BY g.publishedAt DESC
        """
    )
    fun findByTechStacksAndKeyword(
        @Param("techStacks") techStacks: Set<TechStack>,
        @Param("matchedTechStacks") matchedTechStacks: Set<TechStack>,
        @Param("keyword") keyword: String?,
        pageable: Pageable,
    ): Page<GithubRelease>
}